### ISSUE TITLE
Improve/Add new AutoML Sweeping Algorithms

### ISSUE THREAD

**Issue by jpepg4** [dotnet/machinelearning#5608]
### System information

Windows 10
AutoML 0.17.4 
Microsoft.ML 1.5.4

### Issue

When running an experiment AutoML, often after as little as one run, will quickly fixate on 2 trainers until completion.  While this can produce optimal results for those trainers, in ignoring other options it will often miss the best result.  In particular it has a propensity to fixate on the faster logistic regressions vs. slower (but often more effective) SVM, Trees, or GBM trainers. The only workaround seems to be using something like the code below to test two trainers at a time until the best one is found.
experimentSettings.Trainers.Clear();
experimentSettings.Trainers.Add(BinaryClassificationTrainer.LinearSvm);

**Comment by michaelgsharp** [comment]
@justinormont are you familiar with how AutoML picks the trainers it uses?

**Comment by justinormont** [comment]
Yes, I setup the sweeping pattern for AutoML.

## AutoML sweeping pattern

We are using SMAC for hyperparameter optimization. We are optimizing three trainers in parallel, each has a warm-up of [20 random iterations](https://github.com/dotnet/machinelearning/blob/5dbfd8acac0bf798957eea122f1413209cdf07dc/src/Microsoft.ML.AutoML/Sweepers/SmacSweeper.cs#L59), followed by the Bayesian optimization.

Basic sweep/search strategy:
* Test default hyperparameters on each trainer for the task (power of defaults) -- iteration 1 to ~10
  * Initial ordering of the trainers we try is based on speed and general accuracy of the trainers across a variety of datasets, incase time expires before finishing all trainers; also allows ties to sway to the generally faster/better trainer
* _Cull to top three trainers_ ([code](https://github.com/dotnet/machinelearning/blob/796157576fecf2b4b20c47f9451ee88cc428fdaf/src/Microsoft.ML.AutoML/PipelineSuggesters/PipelineSuggester.cs#L107))
* Hyperparameter optimization
  * Random sweeping as warm-up for SMAC -- iteration ~11 to ~70 (3 trainers * 20 iterations)
  * SMAC --  iteration 71+ (until time expires)
  
The culling to the top three trainers is a greedy optimization method allowing us to focus more sweeping time on the candidates which are more likely to be successful.

**Comment by michaelgsharp** [comment]
So assuming time isn't an issue, it will try all the trainers for 10 iterations and then only keep the top 3. Am I understanding that correctly?

@jpepg4, what is the time limit you are using for your runs?

**Comment by justinormont** [comment]
Yes, the first step is **one iteration for each of the trainers** in the task. This uses the trainers' default hyperparameters. 

Each task has a different count of available trainers -- (Binary: 9, Multiclass: 10, Regression: 8, Ranking: 2, Recommendation: 1) -- and the default hyperparameters will be tried for each. This count determines the number of iterations in this first step. 

Then as you say, it keeps the top three trainers.

**Comment by jpepg4** [comment]
So I have been using anywhere from 180 seconds on.  It always makes it through step one but then locks on the three (I believe I said two above but looking again today its because it was using the averaged perception and then two logistic regression trainers).

**Comment by justinormont** [comment]
The focusing on three trainers is expected behavior (not a bug). 

That said, there's always room for improvement. Exploring other methods of sweeping has always been on my list, but wasn't a priority during the time it was being actively worked on. 

Q: How much improvement did you see from other trainers (besides the chosen three)?

If you want to explore each of the trainers, you can run AutoML N times, either with one trainer per run, or up to three. Running independently, allows you to set the time per trainer. If you run three per run, a slow trainer can take most of the runtime.

**Comment by jpepg4** [comment]
> The focusing on three trainers is expected behavior (not a bug).
> 
> That said, there's always room for improvement. Exploring other methods of sweeping has always been on my list, but wasn't a priority during the time it was being actively worked on.
> 
> Q: How much improvement did you see from other trainers (besides the chosen three)?
> 
> If you want to explore each of the trainers, you can run AutoML N times, either with one trainer per run, or up to three. Running independently, allows you to set the time per trainer. If you run three per run, a slow trainer can take most of the runtime.

4 pts of accuracy which may not seem like a lot but this is for a medical application (feel free to e-mail me for details) where its critically important to be as high as possible provided the model is still generalizable.

To your suggestion that's ultimately what I have been doing.  I also found that the model builder didn't recognize a binary classification as binary, and had no option to force that behavior, so I ended up writing a new program, based off the binary classification example, where I can turn on and off a given trainer.  This also allowed me to adapt a secondary generalization test.

**Comment by justinormont** [comment]
Thanks for the datapoint. To clarify is that a 4% gain in accuracy? 

Is that the validation accuracy (as reported by AutoML) or evaluated on a test dataset? During the sweeping process, the validation dataset gets overfit and a holdout/test dataset will give a more useful metric value.

Further gains can be had by using custom featurization, which you then add a `preFeaturizer` to the AutoML API.

**Comment by jpepg4** [comment]
Gain in accuracy on the holdout dataset

**Comment by michaelgsharp** [comment]
@justinormont So I think the actionable issue here is the same as #5609 to just add the other trainers? Or is this more about different sweeping algorithms?

**Comment by justinormont** [comment]
The mentioned `LinearSvm` is already part of AutoML. 

This issue is a request improve the sweeping algorithm(s). Specifically he found `LinearSvm` would have been the winner if the sweeper spent more time optimizing it.

**To add new new AutoML sweeper:**
The AutoML codebase is designed to accommodate multiple sweeping engines. We'd add another sweeper in the [pipeline suggester folder](https://github.com/dotnet/machinelearning/blob/796157576fecf2b4b20c47f9451ee88cc428fdaf/src/Microsoft.ML.AutoML/PipelineSuggesters/). The current one was previously named "rocket engine", and is now just "pipelinesuggester".

**Comment by luisquintanilla** [comment]
Thanks everyone for the discussion and feedback. We've recently updated our implementation of AutoML based on FLAML which supports the requirements listed here.

For more information, see the [blog post](https://devblogs.microsoft.com/dotnet/whats-new-with-mldotnet-automl/#automl-updates) and [sample notebooks](https://github.com/dotnet/csharp-notebooks/blob/main/machine-learning/E2E-Forecasting%20using%20SSA%20with%20Luna%20Dataset.ipynb)

Closing this issue.
